### PR TITLE
bpo-41559: Convert lists to tuples in genericalias to prepare for PEP 612

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Needs manual backport to 3.9.

https://bugs.python.org/issue41559

### CODE REVIEW

File: Objects/genericaliasobject.c [hunk 1]
```diff
@@ -225,6 +226,33 @@ tuple_add(PyObject *self, Py_ssize_t len, PyObject *item)
...
+static inline void
+tupleify_lists(PyObject *args) {
+    Py_ssize_t len = PyTuple_GET_SIZE(args);
+    for (Py_ssize_t i = 0; i < len; ++i) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        int is_list = PyList_CheckExact(arg);
+        int is_tuple = PyTuple_CheckExact(arg);
+        if (is_list || is_tuple) {
+            // In case there are lists nested inside tuples.
+            PyObject *new_arg = is_list ? PyList_AsTuple(arg) : tuple_copy(arg);
```
Comment (ZackerySpytz) [inline Objects/genericaliasobject.c]: You need to check if `new_arg` is `NULL`. Both `PyList_AsTuple()` and `tuple_copy()` can fail and return `NULL`.

File: Objects/genericaliasobject.c [hunk 2]
```diff
@@ -626,10 +641,13 @@ setup_ga(gaobject *alias, PyObject *origin, PyObject *args) {
     else {
         Py_INCREF(args);
     }
+    PyObject *new_args = tuple_copy(args);
+    tupleify_lists(new_args);
```
Comment (ZackerySpytz) [inline Objects/genericaliasobject.c]: Both of these calls should be checked for failure.

File: Objects/genericaliasobject.c [hunk 3]
```diff
@@ -225,6 +226,33 @@ tuple_add(PyObject *self, Py_ssize_t len, PyObject *item)
     return 0;
 }
 
+// Makes a shallow copy of a tuple
+static inline PyObject *
+tuple_copy(PyObject *obj) {
+    return _PyTuple_FromArray(((PyTupleObject *)obj)->ob_item, Py_SIZE(obj));
```
Comment (ZackerySpytz) [inline Objects/genericaliasobject.c]: PEP 7 states that the `{` for function definitions should go on its own line (https://www.python.org/dev/peps/pep-0007/#code-lay-out).

File: Objects/genericaliasobject.c (hunk 1 above)
Comment (ZackerySpytz) [inline Objects/genericaliasobject.c]: The PEP 7 style guide states that lines should be limited to 79 characters.

File: Objects/genericaliasobject.c [hunk 4]
```diff
@@ -225,6 +226,33 @@ tuple_add(PyObject *self, Py_ssize_t len, PyObject *item)
...
+tupleify_lists(PyObject *args) {
+    Py_ssize_t len = PyTuple_GET_SIZE(args);
+    for (Py_ssize_t i = 0; i < len; ++i) {
+        PyObject *arg = PyTuple_GET_ITEM(args, i);
+        int is_list = PyList_CheckExact(arg);
+        int is_tuple = PyTuple_CheckExact(arg);
+        if (is_list || is_tuple) {
+            // In case there are lists nested inside tuples.
+            PyObject *new_arg = is_list ? PyList_AsTuple(arg) : tuple_copy(arg);
+            tupleify_lists(new_arg);
```
Comment (ZackerySpytz) [inline Objects/genericaliasobject.c]: `tupleify_lists()` should not return `void`.

**Comment by Fidget-Spinner** [comment]
@ZackerySpytz Thank you for taking the time to review this and for raising many glaring points :). I'm quite busy right now so I'll try to address them by this weekend.

File: Objects/genericaliasobject.c (hunk 1 above)
Comment (Fidget-Spinner) [reply Objects/genericaliasobject.c]: I'd like to argue against this, PEP 7 says 

> Note, rules are there to be broken. Two good reasons to break a particular rule:
When applying the rule would make the code less readable, even for someone who is used to reading code that follows the rules.

There are many other instances in ``genericaliasobject.c`` which exceed 79 characters (one even up to 91!).
The current line is 81 characters, I'd think a leeway of 2 characters should be afforded :), breaking it into a multi-line ternary would hurt readability IMO as it'd be better off as an if-else.

Comment (Fidget-Spinner) [reply Objects/genericaliasobject.c]: After changing the code a bit, it no longer exceeds 79 chars, so I'm marking this as resolved :).

**Comment by Fidget-Spinner** [comment]
Closed in favour of #25282 which uses a pure Python version of PEP 612. Thus we won't need a C implementation any more.

**Pull request closed without merging**